pkg/syncer/syncer.go

```diff
@@ -67,32 +68,48 @@ func (c *Controller) handleErr(err error, i interface{}) {
 	// Re-enqueue up to 5 times.
 	num := c.Queue.NumRequeues(i)
 	if num < 5 {
-		log.Printf("Error reconciling key %q, retrying... (#%d): %v", i, num, err)
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", i, num, err)
 		c.Queue.AddRateLimited(i)
 		return
 	}
 
 	// Give up and report error elsewhere.
 	c.Queue.Forget(i)
 	utilruntime.HandleError(err)
-	log.Printf("Dropping key %q after failed retries: %v", i, err)
+	klog.Errorf("Dropping key %q after failed retries: %v", i, err)
 }
 
 func (c *Controller) process(gvr schema.GroupVersionResource, obj interface{}) error {
+	klog.Infof("Process object of type: %T : %v", obj, obj)
 	obj, exists, err := c.FromDSIF.ForResource(gvr).Informer().GetIndexer().Get(obj)
 	if err != nil {
+		klog.Error(err)
 		return err
 	}
+	/*
+		unstrob, err := interfaceToUnstructured(obj)
+		if err != nil {
+			klog.Error(err)
+			return err
+		}
+	*/
+	unstrob, isUnstructured := obj.(*unstructured.Unstructured)
 
-	unstrob, err := interfaceToUnstructured(obj)
-	if err != nil {
+	if !isUnstructured {
+		err := errors.New("Object to synchronize is expected to be Unstructured")
+		klog.Error(err)
```

Comment:
Instead of `klog.Error`ing before each `return err`, can we just rely on a single `klog.Error` in handling the error returned by `process`? It looks like we already have that above, so we shouldn't even need these, unless something is slipping through?

Reply:
One big thing with `klog` is that you have the file / line in the code where the error occurred (well, not really... in fact when the  error was *reported*). And that's very handy to understand an error when you cannot debug and have to only read the code to find the error.
Initially, since error was only logged in `handlerErr()` at the end of the processing function, there was no way to understand where precisely the error was triggered.
I'd rather log errors (with the detailed message) as near as possible from where they occur.

Since we're doing a prototype here, it doesn't seem a problem to me if logs are a bit too wordy, especially if it's only to provide more details in cases of error.

Reply:
I understand. Can we move the klog change to a separate, smaller, focused PR, where it's applied evenly across the codebase?

A pattern I've seen is that when `if err != nil { return err }` is replaced with `if err != nil { log.Error(err); return err }` is that it's easy to forget to add the `log` part consistently, and that weakens its usefulness. This isn't `klog`'s fault or `zap`'s or anybody's, except maybe the language for not having better support for this.

An alternative (not my prescribed mandated answer, just an option) is to `return fmt.Errorf("failed trying to 'blah': %w", err)`, which adds debugging information and is harder to miss (and is a line shorter).

Anyway, having this discussion in a separate PR focused on how we want to return and bubble up error information would let us have this conversation separate from the rest of the PR, which we shouldn't otherwise block.

Reply:
Would it be possible to merge this one and open an issue about making the log management homogeneous ?
I don't really see the point in removing more detailed logs when people might start trying it en encounter failures...